DiffInYears producing incorrect output

Hi, I've noticed `DiffInYears` seems to produce incorrect output, unless I'm mistaken.

Test case:

``` go
package main

import (
    "testing"
    "time"

    "github.com/stretchr/testify/assert"
    "github.com/uniplaces/carbon"
)

func TestCarbonDiffInYears(t *testing.T) {
    dob, _ := carbon.CreateFromDate(2000, time.June, 27, time.UTC.String())

    yesterday, _ := carbon.CreateFromDate(2016, time.June, 26, time.UTC.String())
    today, _ := carbon.CreateFromDate(2016, time.June, 27, time.UTC.String())
    tomorrow, _ := carbon.CreateFromDate(2016, time.June, 28, time.UTC.String())

    // Day before 16th birthday... Should be 15
    assert.Equal(t, int64(15), dob.DiffInYears(yesterday, true))
    // Day of 16th birthday
    assert.Equal(t, int64(16), dob.DiffInYears(today, true))
    // Day after 16th birthday
    assert.Equal(t, int64(16), dob.DiffInYears(tomorrow, true))
}
```

Result:

```
$ go test -v
=== RUN   TestCarbonDiffInYears
--- FAIL: TestCarbonDiffInYears (0.00s)
        Error Trace:    carbon_test.go:19
    Error:      Not equal:
            expected: 15
            received: 16

FAIL
exit status 1
FAIL    github.com/bbrks/carbon-test    0.011s
```

--- a/carbon.go
+++ b/carbon.go
@@ -872,12 +872,12 @@ func (c *Carbon) IsTomorrow() bool {
 
 // IsFuture determines if the current time is in the future, ie. greater (after) than now
 func (c *Carbon) IsFuture() bool {
-	return c.After(time.Now())
+	return c.After(Now().Time)
 }
 
 // IsPast determines if the current time is in the past, ie. less (before) than now
 func (c *Carbon) IsPast() bool {
-	return c.Before(time.Now())
+	return c.Before(Now().Time)
 }
 
 // IsLeapYear determines if current current time is a leap year
@@ -991,7 +991,7 @@ func (c *Carbon) IsSaturday() bool {
 // IsLastWeek returns true is the date is within last week
 func (c *Carbon) IsLastWeek() bool {
 	secondsInWeek := float64(secondsInWeek)
-	difference := time.Now().Sub(c.Time)
+	difference := Now().Sub(c.Time)
 	if difference.Seconds() > 0 && difference.Seconds() < secondsInWeek {
 		return true
 	}
@@ -1002,7 +1002,7 @@ func (c *Carbon) IsLastWeek() bool {
 // IsLastMonth returns true is the date is within last month
 func (c *Carbon) IsLastMonth() bool {
 	secondsInMonth := float64(secondsInMonth)
-	difference := time.Now().Sub(c.Time)
+	difference := Now().Sub(c.Time)
 	if difference.Seconds() > 0 && difference.Seconds() < secondsInMonth {
 		return true
 	}

--- a/carbon.go
+++ b/carbon.go
@@ -1001,9 +1001,15 @@ func (c *Carbon) IsLastWeek() bool {
 
 // IsLastMonth returns true is the date is within last month
 func (c *Carbon) IsLastMonth() bool {
-	secondsInMonth := float64(secondsInMonth)
-	difference := Now().Sub(c.Time)
-	if difference.Seconds() > 0 && difference.Seconds() < secondsInMonth {
+	now := Now()
+
+	monthDifference := now.Month() - c.Month()
+
+	if absValue(true, int64(monthDifference)) != 1 {
+		return false
+	}
+
+	if now.UnixNano() > c.UnixNano() && monthDifference == 1 {
 		return true
 	}
 

--- a/carbon.go
+++ b/carbon.go
@@ -1152,16 +1152,23 @@ func (c *Carbon) DiffInYears(carb *Carbon, abs bool) int64 {
 		return 0
 	}
 
-	diffHr := c.DiffInHours(carb, abs)
-	hrLastYear := int64(c.DaysInYear() * hoursPerDay)
-
-	if (diffHr - hrLastYear) >= 0 {
-		diff := int64(carb.In(time.UTC).Year() - c.In(time.UTC).Year())
-
-		return absValue(abs, diff)
+	start := NewCarbon(c.Time)
+	end := NewCarbon(carb.Time)
+	if end.UnixNano() < start.UnixNano() {
+		aux := start
+		start = end
+		end = aux
 	}
 
-	return 0
+	yearsAmmount := int64(end.Year()-start.Year()) - 1
+
+	start.SetYear(end.Year())
+
+	if start.UnixNano() <= end.UnixNano() {
+		yearsAmmount++
+	}
+
+	return absValue(abs, yearsAmmount)
 }
 
 // DiffInMonths returns the difference in months

--- a/carbon.go
+++ b/carbon.go
@@ -67,6 +67,12 @@ type Carbon struct {
 	Translator   *Translator
 }
 
+// Used for testing purposes
+var (
+	isTimeFrozen      bool
+	currentFrozenTime time.Time
+)
+
 // NewCarbon returns a pointer to a new Carbon instance
 func NewCarbon(t time.Time) *Carbon {
 	wds := []time.Weekday{
@@ -81,6 +87,64 @@ func NewCarbon(t time.Time) *Carbon {
 		stringFormat: DefaultFormat,
 		Translator:   translator(),
 	}
+}
+
+// Freeze allows time to be frozen to facilitate testing
+func Freeze(time time.Time) {
+	currentFrozenTime = time
+	isTimeFrozen = true
+}
+
+// UnFreeze returns time to normal operation
+func UnFreeze() {
+	isTimeFrozen = false
+}
+
+// IsTimeFrozen allows checking if time has been frozen
+func IsTimeFrozen() bool {
+	return isTimeFrozen
+}
+
+// After will be behave like time.After unless time has been frozen
+// If time is frozen it will add the expected delay and immediately send the frozen time on the returned channel
+func After(d time.Duration) <-chan time.Time {
+	if isTimeFrozen {
+		currentFrozenTime = currentFrozenTime.Add(d)
+		c := make(chan time.Time, 1)
+		c <- currentFrozenTime
+		return c
+	}
+
+	return time.After(d)
+}
+
+// Tick will be behave like time.Tick unless time has been frozen
+// If time is frozen it will tick normally but the date will be based on the frozen date
+func Tick(d time.Duration) <-chan time.Time {
+	if isTimeFrozen {
+		c := make(chan time.Time, 1)
+		go func() {
+			for {
+				currentFrozenTime = currentFrozenTime.Add(d)
+				c <- currentFrozenTime
+			}
+		}()
+		return c
+	}
+
+	return time.Tick(d)
+}
+
+// Sleep will be behave like time.Sleep unless time has been frozen
+// If time is frozen it will add the expected sleep delay and return immediately
+func Sleep(d time.Duration) {
+	if isTimeFrozen && d > 0 {
+		currentFrozenTime = currentFrozenTime.Add(d)
+
+		return
+	}
+
+	time.Sleep(d)
 }
 
 // create returns a new carbon pointe. It is a helper function to create new dates
@@ -216,6 +280,10 @@ func MinValue() *Carbon {
 
 // Now returns a new Carbon instance for right now in current localtime
 func Now() *Carbon {
+	if isTimeFrozen {
+		return NewCarbon(currentFrozenTime)
+	}
+
 	return NewCarbon(time.Now())
 }
 

--- a/carbon_test.go
+++ b/carbon_test.go
@@ -1919,7 +1919,7 @@ func TestDiffInYearsOneHourLessInYear(t *testing.T) {
 
 func TestDiffInYearsOneYearDifferenceForLeapYear(t *testing.T) {
 	t1, _ := Create(2015, time.August, 10, 15, 0, 0, 0, "UTC")
-	t2, _ := Create(2016, time.August, 10, 14, 0, 0, 0, "UTC")
+	t2, _ := Create(2016, time.August, 10, 15, 0, 0, 0, "UTC")
 
 	assert.EqualValues(t, 1, t1.DiffInYears(t2, true))
 }
@@ -2531,9 +2531,10 @@ func TestCreateFromDate(t *testing.T) {
 
 func TestAge(t *testing.T) {
 	c, _ := CreateFromDate(2011, time.August, 10, "UTC")
-	y := Now().Year()
+	y, _ := CreateFromDate(2017, time.August, 10, "UTC")
+	Freeze(y.Time)
 
-	assert.EqualValues(t, y-2011, c.Age())
+	assert.EqualValues(t, 6, c.Age())
 }
 
 func TestWeekOfMonth(t *testing.T) {
@@ -2625,6 +2626,8 @@ func TestFirstDayOfMonth(t *testing.T) {
 }
 
 func TestIsLastWeekTrue(t *testing.T) {
+	t.Skip("IsLastWeek is a little flacky, there is no definition of when a week starts...")
+
 	c := Now().SubDays(3)
 	assert.True(t, c.IsLastWeek())
 }
@@ -2636,12 +2639,18 @@ func TestIsLastWeekFalse(t *testing.T) {
 }
 
 func TestIsLastMonthTrue(t *testing.T) {
-	c := Now().SubWeeks(2)
+	c, _ := Create(2016, time.May, 5, 10, 0, 0, 0, "UTC")
+	Freeze(c.Time)
+
+	c = Now().SubWeeks(2)
 	assert.True(t, c.IsLastMonth())
 }
 
 func TestIsLastMonthFalse(t *testing.T) {
 	c, err := Create(2016, time.May, 20, 10, 0, 0, 0, "UTC")
+	Freeze(c.Time)
+
+	c = Now().SubWeek()
 	assert.Nil(t, err)
 	assert.False(t, c.IsLastMonth())
 }

--- a/carbonfreeze_test.go
+++ b/carbonfreeze_test.go
@@ -1,0 +1,71 @@
+package carbon_test
+
+import (
+	"testing"
+	"time"
+
+	"github.com/uniplaces/carbon"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFreeze(t *testing.T) {
+	sleepDuration := 2 * time.Second
+
+	// 2017-03-08T01:24:34+00:00
+	expectedTimeStamp := int64(1488936274)
+	expectedFrozenSleepTimeStamp := expectedTimeStamp + int64(sleepDuration.Seconds())
+
+	timeToFreeze, err := carbon.CreateFromTimestampUTC(expectedTimeStamp)
+	assert.Nil(t, err)
+
+	carbon.Freeze(timeToFreeze.Time)
+	defer carbon.UnFreeze()
+
+	realNowBefore := time.Now()
+	frozenNowBefore := carbon.Now()
+
+	// Really sleep
+	time.Sleep(sleepDuration)
+
+	realNowAfter := time.Now()
+	frozenNowAfter := carbon.Now()
+
+	assert.NotEqual(t, realNowBefore, realNowAfter)
+	assert.True(t, time.Since(realNowBefore).Seconds() >= sleepDuration.Seconds())
+	assert.Equal(t, frozenNowBefore, frozenNowAfter)
+	assert.Equal(t, expectedTimeStamp, frozenNowBefore.Unix())
+	assert.Equal(t, expectedTimeStamp, frozenNowAfter.Unix())
+
+	carbon.Sleep(sleepDuration)
+
+	frozenAfterSleep := carbon.Now()
+	assert.NotEqual(t, frozenNowAfter, frozenAfterSleep)
+	assert.Equal(t, expectedFrozenSleepTimeStamp, frozenAfterSleep.Unix())
+}
+
+func TestUnFreeze(t *testing.T) {
+	sleepDuration := 1 * time.Second
+
+	// 2017-03-08T01:24:34+00:00
+	expectedTimeStamp := int64(1488936274)
+
+	timeToFreeze, err := carbon.CreateFromTimestampUTC(expectedTimeStamp)
+	assert.Nil(t, err)
+
+	carbon.Freeze(timeToFreeze.Time)
+	defer carbon.UnFreeze()
+
+	frozenNow := carbon.Now()
+
+	// Really sleep
+	time.Sleep(sleepDuration)
+
+	carbon.UnFreeze()
+
+	realNow := time.Now()
+	unfrozenNow := carbon.Now()
+
+	assert.NotEqual(t, frozenNow, unfrozenNow)
+	assert.Equal(t, expectedTimeStamp, frozenNow.Unix())
+	assert.Equal(t, realNow.Unix(), unfrozenNow.Time.Unix())
+}

--- a/carbonfreeze_test.go
+++ b/carbonfreeze_test.go
@@ -4,8 +4,8 @@ import (
 	"testing"
 	"time"
 
-	"github.com/uniplaces/carbon"
 	"github.com/stretchr/testify/assert"
+	"github.com/uniplaces/carbon"
 )
 
 func TestFreeze(t *testing.T) {
@@ -68,4 +68,30 @@ func TestUnFreeze(t *testing.T) {
 	assert.NotEqual(t, frozenNow, unfrozenNow)
 	assert.Equal(t, expectedTimeStamp, frozenNow.Unix())
 	assert.Equal(t, realNow.Unix(), unfrozenNow.Time.Unix())
+}
+
+func TestNotSharing(t *testing.T) {
+	// 2017-03-08T01:24:34+00:00
+	expectedTimeStamp1 := int64(1488936274)
+	timeToFreeze1, _ := carbon.CreateFromTimestampUTC(expectedTimeStamp1)
+	carbon.Freeze(timeToFreeze1.Time)
+
+	carbon1 := carbon.Now()
+
+	time.Sleep(2 * time.Second)
+
+	carbon11 := carbon.Now()
+
+	// 2017-03-09T10:12:01+00:00
+	expectedTimeStamp2 := int64(1489054321)
+	timeToFreeze2, _ := carbon.CreateFromTimestampUTC(expectedTimeStamp2)
+	carbon.Freeze(timeToFreeze2.Time)
+
+	carbon2 := carbon.Now()
+
+	assert.NotEqual(t, carbon1.UnixNano(), carbon2.UnixNano())
+	assert.Equal(t, carbon1.UnixNano(), carbon11.UnixNano())
+	assert.Equal(t, expectedTimeStamp1, carbon1.Unix())
+	assert.Equal(t, expectedTimeStamp1, carbon11.Unix())
+	assert.Equal(t, expectedTimeStamp2, carbon2.Unix())
 }

--- a/diff_test.go
+++ b/diff_test.go
@@ -2,9 +2,26 @@ package carbon
 
 import (
 	"testing"
+	"time"
 
 	"github.com/stretchr/testify/assert"
 )
+
+// Check https://github.com/uniplaces/carbon/issues/30
+func TestCarbonDiffInYears(t *testing.T) {
+	dob, _ := CreateFromDate(2000, time.June, 27, time.UTC.String())
+
+	yesterday, _ := CreateFromDate(2016, time.June, 26, time.UTC.String())
+	today, _ := CreateFromDate(2016, time.June, 27, time.UTC.String())
+	tomorrow, _ := CreateFromDate(2016, time.June, 28, time.UTC.String())
+
+	// Day before 16th birthday... Should be 15
+	assert.Equal(t, int64(15), dob.DiffInYears(yesterday, true))
+	// Day of 16th birthday
+	assert.Equal(t, int64(16), dob.DiffInYears(today, true))
+	// Day after 16th birthday
+	assert.Equal(t, int64(16), dob.DiffInYears(tomorrow, true))
+}
 
 func TestDiffForHumansNowAndSecond(t *testing.T) {
 	gotTime, err := Now().DiffForHumans(nil, false, false, false)

